ui/StatusQ/src/functionaggregator.cpp

```diff
@@ -0,0 +1,98 @@
+#include "StatusQ/functionaggregator.h"
+
+#include <QDebug>
+#include <QJSEngine>
+
+FunctionAggregator::FunctionAggregator(QObject* parent)
+    : SingleRoleAggregator(parent)
+{
+    recalculate();
+}
+
+const QVariant& FunctionAggregator::initialValue() const
+{
+    return m_initialValue;
+}
+
+void FunctionAggregator::setInitialValue(const QVariant& initialValue)
+{
+    if (m_initialValue == initialValue)
+        return;
+
+    m_initialValue = initialValue;
+
+    emit initialValueChanged();
+    recalculate();
+}
+
+const QJSValue& FunctionAggregator::aggregateFunction() const
+{
+    return m_aggregateFunction;
+}
+
+void FunctionAggregator::setAggregateFunction(const QJSValue& aggregateFunction)
+{
+    if (m_aggregateFunction.strictlyEquals(aggregateFunction))
+        return;
+
+    if (!aggregateFunction.isCallable() && !aggregateFunction.isUndefined()) {
+        qWarning() << "FunctionAggregator::aggregateFunction must be a "
+                      "callable object.";
+        return;
+    }
+
+    m_aggregateFunction = aggregateFunction;
+
+    emit aggregateFunctionChanged();
+    recalculate();
+}
+
+QVariant FunctionAggregator::calculateAggregation()
+{
+    // Check if m_model exists and role name is initialized
+    if (!model() || roleName().isEmpty())
+        return m_initialValue;
+
+    // Check if m_roleName is part of the roles of the model
+    QHash<int, QByteArray> roles = model()->roleNames();
+    if (!roleExists() && model()->rowCount()) {
+        qWarning() << "Provided role name does not exist in the current model.";
+        return m_initialValue;
+    }
+
+    if (!m_initialValue.isValid())
```

Comment:
I don't understand this condition

Reply:
Initial value is for example `0` when we calculate sum or empty array `[]` when we want to gather some role values as a list. When it's not provided, it's just `undefined` (invalid variant), we have no starting point for calculating that aggregation. So we immediately return that undefined as a result instead of passing it to the function.

Reply:
Got it!